src/rust/catloop/mod.rs

```diff
@@ -0,0 +1,521 @@
+            Some((demi_opcode_t::DEMI_OPC_ACCEPT, _)) | Some((demi_opcode_t::DEMI_OPC_CONNECT, _)) => {
+                let (qd, r): (QDesc, OperationResult) = self.take_result(handle);
+
+                match r {
+                    OperationResult::Connect((remote, ref duplex_pipe)) => {
+                        _ = self.sockets.insert(qd, Some((remote, duplex_pipe.clone())))
+                    },
+                    OperationResult::Accept((new_qd, (remote, ref duplex_pipe))) => {
+                        self.sockets.insert(new_qd, Some((remote, duplex_pipe.clone())));
+                    },
+                    _ => {},
+                };
+
+                return Ok(pack_result(r, qd, qt.into()));
+            },
+            Some((demi_opcode_t::DEMI_OPC_PUSH, qd)) | Some((demi_opcode_t::DEMI_OPC_POP, qd)) => {
+                let mut qr: demi_qresult_t = self.catmem.borrow_mut().pack_result(handle, qt)?;
+                qr.qr_qd = qd.into();
+
+                if qr.qr_opcode == demi_opcode_t::DEMI_OPC_POP {
+                    let mut sga: &mut demi_sgarray_t = unsafe { &mut qr.qr_value.sga };
+
+                    // Check for close.
+                    if self.is_magic_close(sga) {
+                        sga.sga_segs[0].sgaseg_len = 0;
+                    }
+                }
+
+                return Ok(qr);
+            },
+            Some((demi_opcode_t::DEMI_OPC_FAILED, qd)) => {
+                // ToDo: handle failure correctly. If an accept() operation failed, rollback port allocation.
+                let message: String = format!("operation failed (qd={:?}", qd);
+                let e: Fail = Fail::new(libc::EAGAIN, &message);
+                return Err(e);
+            },
+            _ => return Err(Fail::new(libc::EINVAL, "invalid queue token")),
+        }
+    }
+
+    /// Polls scheduling queues.
+    pub fn poll(&self) {
+        self.catmem.borrow().poll();
+        self.scheduler.poll()
+    }
+
+    /// Takes out the [OperationResult] associated with the target [SchedulerHandle].
+    fn take_result(&mut self, handle: SchedulerHandle) -> (QDesc, OperationResult) {
+        let boxed_future: Box<dyn Any> = self.scheduler.take(handle).as_any();
+        let boxed_concrete_type: Operation = *boxed_future.downcast::<Operation>().expect("Wrong type!");
+
+        boxed_concrete_type.get_result()
+    }
+
+    // Cooks a magic connect message.
+    pub fn cook_magic_connect(catmem: &Rc<RefCell<CatmemLibOS>>) -> Result<demi_sgarray_t, Fail> {
+        let sga: demi_sgarray_t = catmem
+            .borrow_mut()
+            .alloc_sgarray(mem::size_of_val(&CatloopLibOS::MAGIC_CONNECT))?;
+
+        let ptr: *mut u8 = sga.sga_segs[0].sgaseg_buf as *mut u8;
+        let len: usize = sga.sga_segs[0].sgaseg_len as usize;
+        let slice: &mut [u8] = unsafe { slice::from_raw_parts_mut(ptr, len) };
+        slice.copy_from_slice(&CatloopLibOS::MAGIC_CONNECT.to_ne_bytes());
```

Comment:
Since MAGIC_CONNECT is just a single byte, why all this code in this function (and `is_magic_connect`) to create a slice?

Reply:
Oh, thanks! I applied the optimization that you suggested.